### PR TITLE
Added traversals and shortest path algos

### PR DESCRIPTION
**Pull request by npochhi**
### Description

Added graph-traversals and shortest path algorithms

- Relevant Issues : (required)
- Relevant PRs : (optional)
- Type of change :
  - [x] New feature
  - [ ] Bug fix for existing feature
  - [ ] Code quality improvement
  - [ ] Addition or Improvement of tests
  - [ ] Addition or Improvement of documentation

### CODE REVIEW

File: lib/networkx/graph.rb [hunk 1]
```diff
@@ -325,5 +325,13 @@ def edge_subgraph(edges)
...
       end
     end
+
+    def is_multigraph
+      self.class.name.include?('Multi')
+    end
+
+    def is_directed
+      self.class.name.include?('Di')
+    end
```
Comment (athityakumar) [inline lib/networkx/graph.rb]: This works, but why not have something more "safer" (not open-ended) like,

```ruby
[NetworkX::DiGraph, NetworkX::MultiDiGraph].include?(self.class)
```

File: lib/networkx/graph.rb [hunk 2]
```diff
@@ -325,5 +325,13 @@ def edge_subgraph(edges)
         "received #{edges.class.name} instead."
       end
     end
+
+    def is_multigraph
+      self.class.name.include?('Multi')
+    end
+
+    def is_directed
```
Comment (athityakumar) [inline lib/networkx/graph.rb]: Also, it'd be a more Rubyistic nomenclature to rename these methods with a `?` like `is_multigraph?` and `is_directed?`.

File: lib/networkx/shortest_path/astar.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
+require 'rb_heap'
+require 'pry'
```
Comment (athityakumar) [inline lib/networkx/shortest_path/astar.rb]: Why has `pry` been `require`d? 🤔

File: lib/networkx/shortest_path/weighted.rb [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  end
+
+  def self.allpairs_bellmanford_path(g, cutoff=nil)
+    paths = []
+    g.nodes.each_key { |n| paths << [n, singlesource_bellmanford_path(g, n, cutoff)] }
+    paths
+  end
+
+  def self.goldberg_radzik(g, source)
+    raise NotImplementedError, 'Not yet implemented!'
```
Comment (athityakumar) [inline lib/networkx/shortest_path/weighted.rb]: Implement this before the next review of this PR.

File: networkx.gemspec [hunk 5]
```diff
@@ -21,7 +21,9 @@ Gem::Specification.new do |spec|
   spec.require_paths = ['lib']
 
   spec.add_development_dependency 'bundler', '~> 1.15'
+  spec.add_development_dependency 'nmatrix'
```
Comment (athityakumar) [inline networkx.gemspec]: `nmatrix` and `rb_heap` are runtime dependencies, not development dependencies

File: spec/network_x/traversals/dfs_spec.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+require 'spec_helper'
```
Comment (athityakumar) [inline spec/network_x/traversals/dfs_spec.rb]: These `spec_helper` requires are absolutely not required. Try running the specs without them. 😄

File: spec/network_x/traversals/edge_dfs_spec.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    graph.add_edges([%w[B A], %w[B D]])
+    graph.add_node('E')
+  end
+
+  context 'when dfs_edges is called' do
+    subject { NetworkX.edge_dfs(graph, 'A') }
+
+    it do
+      is_expected.to eq([["A", "B"], ["B", "A"], ["B", "D"]])
+    end
```
Comment (athityakumar) [inline spec/network_x/traversals/edge_dfs_spec.rb]: It kinda feels weird to read an `it do` condition without any description (I know that's been covered by the context). `it { is_expected.to eq(something)}` reads more naturally than a `do-end` block.

Has this been done for rubocop?

Comment (npochhi) [reply spec/network_x/traversals/edge_dfs_spec.rb]: No, I haven't done any offense removal till now. Will do it shortly.

File: networkx.gemspec (hunk 5 above)
Comment (npochhi) [reply networkx.gemspec]: Ah, I'm not well versed with gemspec. Sorry for this.. :slightly_frowning_face:

File: lib/networkx/graph.rb (hunk 2 above)
Comment (npochhi) [reply lib/networkx/graph.rb]: I saw an offense related to these two. Will modify it shortly.

File: lib/networkx/graph.rb (hunk 1 above)
Comment (npochhi) [reply lib/networkx/graph.rb]: Sure.

File: spec/network_x/traversals/dfs_spec.rb (hunk 6 above)
Comment (npochhi) [reply spec/network_x/traversals/dfs_spec.rb]: Okay, will remove it. Also, is it necessary to write specs for each and every function? I haven't written specs for some straightforward functions, like the ones which calculate path lengths. If the paths calculated by them are correct, the path lengths would be correct too.

**Pull request merged**